sdk/core/azure-core/src/main/java/com/azure/core/util/ProgressListener.java

```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+/**
+ * A {@link ProgressListener} is an interface that can be used to report progress on network transfers.
+ * The {@link #onProgress(long)} method will be called periodically with the total number of bytes
+ * transferred.
+ */
+public interface ProgressListener {
+    /**
+     * The callback function invoked as progress is reported.
+     *
+     * <p>
+     * The callback can be called concurrently from multiple threads if reporting spans across multiple
+     * requests.
+     * </p>
+     *
+     * @param bytesTransferred The total number of bytes transferred during this transaction.
+     */
+    void onProgress(long bytesTransferred);
```

Comment:
This could have an `@FunctionalInterface` annotation (at class level). I also wonder if we should be reporting the total size, if known, so that users can more easily determine percentage progress?

Reply:
It's replacing https://github.com/Azure/azure-sdk-for-java/search?q=%22interface+ProgressReceiver%22 .

I'll add `@FunctionalInterface`

I wouldn't add total size eagerly. We have APIs (e.g. in shares) that take Fluxes or InputStreams without defined length and chunk them.
If caller knows the length they can easily implement
```
class MyReporter {
 private final long totalSize;
 public MyReporter(long totalsize);
 public onProgress(long delta) {
 System.out.println(delta/totalSize + " completed);
}
}

shareClient.upload(flux, new MyReporter(knownSize));
```

If the caller wouldn't know total size upfront, we wouldn't know it either.

Reply:
After conducting this experiment https://github.com/Azure/azure-sdk-for-java/pull/29277/commits/36b58fde61222042d7915b88a92313633aacb819 . I think we should:
- match signature of existing `ProgressReceiver` in storage and comms
- eventually make `storage/comms.ProgressReceiver` extend `core.ProgressReceiver` so that users with existing implementations will be able to use it with new apis where we'd use core's type.

Reply:
https://github.com/kasobol-msft/azure-sdk-for-java/commit/9f37baef0bbc4da0396af3a5f05f0ead89f11095